server/bin/pbench-tarball-stats

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+"""Pbench Tar Ball Stats
+
+Scan through the ARCHIVE hierarchy for tar balls and report some
+statistics on them:
+
+  * Total number of controllers
+  * Tar balls per controller (most to least)
+  * Tar balls by month for all years, count
+  * Tar balls by month for all years, total size
+  * Top 10 most recent controllers
+  * Top 10 largest tar balls
+"""
+
+import collections
+import locale
+import os
+import re
+import sys
+
+from datetime import datetime
+from argparse import ArgumentParser
+
+import pbench
+from pbench import (
+    PbenchConfig,
+    BadConfig,
+    get_pbench_logger,
+)
+
+
+_NAME_ = "pbench-tarball-stats"
+
+tb_pat = re.compile(
+    r"\S+_(\d\d\d\d)[._-](\d\d)[._-](\d\d)[T_](\d\d)[._:](\d\d)[._:](\d\d)\.tar\.xz"
+)
+
+
+def gen_tar_balls(archive):
+    """Traverse the given ARCHIVE hierarchy looking tar balls, returning
+    information about the tar ball so that statistics can be generated.
+    """
+    with os.scandir(archive) as archive_scan:
+        for c_entry in archive_scan:
+            if c_entry.name.startswith(".") or not c_entry.is_dir(follow_symlinks=False):
+                continue
+            # We have a controller directory.
+            with os.scandir(c_entry.path) as controller_scan:
+                for entry in controller_scan:
+                    if entry.name.startswith(".") or not entry.is_file():
+                        continue
+                    match = tb_pat.fullmatch(entry.name)
+                    if match:
+                        # Turn the pattern components of the match into a datetime
+                        # object.
+                        tb_dt = datetime(
+                            int(match.group(1)),
+                            int(match.group(2)),
+                            int(match.group(3)),
+                            int(match.group(4)),
+                            int(match.group(5)),
+                            int(match.group(6)),
+                        )
+                    else:
+                        if entry.name.endswith(".md5"):
+                            continue
+                        tb_dt = None
```

Comment:
If we get here, it's because the file is "invalid" (so long as it's not an MD5 file), and you're signaling that to the caller by setting `tb_dt` to `None`...a comment (here and in the docstring?) explaining this would go a long way.

(Also, I would vote for using an `elif` for the MD5 check and an `else` for the `tb_dt = None`, since you really do have three co-equal cases.)

(And, I would have been tempted, when I hit an MD5 file, to go confirm that the corresponding tar file exists, and to declare the MD5 file "invalid" if the tarball is missing...but, perhaps that is beyond the scope of this tool.)